File: shortest_unsorted_continuous_subarray.py
class Solution:
    def findUnsortedSubarray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        start = 0
        end = len(nums)
        for i in range(len(nums)):
            if i + 1 < len(nums) and nums[i] > nums[i + 1]:
                start = i
                break
        
        for i in range(len(nums) - 1, -1, -1):
            if i - 1 >= 0 and nums[i] < nums[i - 1]:
                end = i
                break
        
        return (end - start + 1)

File: shortest_unsorted_continuous_subarray.js
const findPosition = (stack, num, nums, position) => {
    const latest = stack.pop()
    if (position == "start" && nums[latest] <= num) return (latest + 1)
    if (position == "end" && nums[latest] >= num) return (latest - 1)
    findPosition(stack, num, nums)
}

const findUnsortedSubarray = nums => {
    let firstPosition = 0 
    let secondPosition = 0
    let stack = []
    let pos = 0
    for (num of nums) {
        if (stack && nums[stack[-1]] > num) {
            firstPosition = findPosition(stack, num, nums, "start")
            break
        }
        stack.push(pos)
        pos++
    }

    stack = []
    pos = nums.length - 1

    for (num of nums.reverse()) {
        if (stack && nums[stack[-1]] < num) {
            secondPosition = findPosition(stack, num, nums, "end")
            break
        }
        stack.push(pos)
        pos--
    }

    return (secondPosition - firstPosition)
}